parsec/core/local_storage.py

```diff
@@ -261,3 +263,106 @@ def create_cursor(self, entry_id: EntryID) -> FileDescriptor:
         cursor = FileCursor(entry_id)
         self.add_file_reference(entry_id)
         return self.create_file_descriptor(cursor)
+
+    def to_timestamped(self, timestamp: Pendulum):
+        return LocalStorageTimestamped(self, timestamp)
+
+
+class LocalStorageTimestamped(LocalStorage):
+    """Timestamped version to access a local storage as it was at a given timestamp
+
+    That includes:
+    - another cache in memory for fast access to deserialized data
+    - the timestamped persistent storage to keep serialized data on the disk :
+      vlobs are in common, not manifests. Actually only vlobs are used, manifests are mocked
+    - a lock mecanism to protect against race conditions OR
+    - a mocked lock mecanism to stay coherent with the actual LocalStorage
+    """
+
+    def __init__(self, local_storage: LocalStorage, timestamp: Pendulum):
+        self.device_id = local_storage.device_id
+
+        # Cursors and file descriptors
+        self.open_cursors: Dict[FileDescriptor, FileCursor] = {}
+        self.file_references: Dict[EntryID, Set[Union[EntryID, FileDescriptor]]] = defaultdict(set)
+        self._fd_counter = 0
+
+        # Locking structures
+        self.locking_tasks = {}
+        self.entry_locks = defaultdict(trio.Lock)
+
+        # Manifest and block storage
+        self.base_manifest_cache = {}
+        self.local_manifest_cache = {}  # should delete? seems used for dirty, base for clean...
+
+        self.persistent_storage = local_storage.persistent_storage
+
+    # Locking helpers
+
+    @asynccontextmanager
+    async def lock_entry_id(self, entry_id: EntryID):
+        async with self.entry_locks[entry_id]:
+            try:
+                self.locking_tasks[entry_id] = hazmat.current_task()
+                yield entry_id
+            finally:
+                del self.locking_tasks[entry_id]
+
+    @asynccontextmanager
+    async def lock_manifest(self, entry_id: EntryID):
+        async with self.lock_entry_id(entry_id):
+            yield self.get_manifest(entry_id)
+
+    def _check_lock_status(self, entry_id: EntryID) -> None:
+        task = self.locking_tasks.get(entry_id)
+        # TODO: remove `task is None` to ensure that the lock is taken
+        assert task is None or task == hazmat.current_task()
+
+    # Manifest interface
+
+    def get_base_manifest(self, entry_id: EntryID) -> RemoteManifest:
+        """Raises: LocalStorageMissingError"""
+        assert isinstance(entry_id, EntryID)
+        try:
+            return self.base_manifest_cache[entry_id]
+        except KeyError:
+            raise LocalStorageMissingError(entry_id)
+
+    def get_manifest(self, entry_id: EntryID) -> LocalManifest:
+        """Raises: LocalStorageMissingError"""
+        assert isinstance(entry_id, EntryID)
+        try:
+            return self.local_manifest_cache[entry_id]
+        except KeyError:
+            raise LocalStorageMissingError
```

Comment:
`get_manifest` may also return a local manifest from the remote manifest in `base_manifest_cache` if it exists